### PR TITLE
ENG-5989 | Mendeley API Implementation for CitationsInterface in GV

### PR DESCRIPTION
**Pull request by sh-andriy**
Implementation of **Mendeley API** for `CitationsInterface` for `GV`. More Details in the ticket (`ENG-5989`)

### CODE REVIEW

File: addon_toolkit/interfaces/citation.py [hunk 1]
```diff
@@ -38,17 +38,17 @@ class ItemType(Enum):
 
 @dataclasses.dataclass(slots=True)
 class ItemResult:
-    item_id: int
+    item_id: str
     item_name: str
-    item_type: ItemType
+    item_type: str
```
Comment (adlius) [inline addon_toolkit/interfaces/citation.py:43]: why is `item_type` not of type `ItemType` ?

File: addon_imps/citations/mendeley.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        async with self.network.GET(
+            "profiles/me",
+        ) as response:
+            profile_info = await response.json_content()
+            user_id = profile_info.get("id")
+            if not user_id:
+                raise KeyError("Failed to fetch user ID from Mendeley.")
+            return str(user_id)
+
+    async def list_root_collections(self) -> ItemSampleResult:
```
Comment (adlius) [inline addon_imps/citations/mendeley.py:22]: I think we may have forgotten about paginations.

File: addon_imps/citations/mendeley.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        async with self.network.GET("folders") as response:
+            response_json = await response.json_content()
+            return self._parse_collection_response(response_json)
+
+    async def list_collection_items(
```
Comment (adlius) [inline addon_imps/citations/mendeley.py:27]: Same issues with paginations here.

**Review by adlius — CHANGES_REQUESTED** [review]
First pass done.

File: addon_imps/citations/mendeley.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            f"folders/{collection_id}/documents",
+        ) as response:
+            import pydevd_pycharm
+
+            pydevd_pycharm.settrace(
+                "host.docker.internal",
+                port=12346,
+                stdoutToServer=True,
+                stderrToServer=True,
+            )
```
[the remark this answers is not in the document]
Comment (sh-andriy) [reply addon_imps/citations/mendeley.py:42]: oh yea, didn't see that one. Will remove it

File: addon_toolkit/interfaces/citation.py (hunk 1 above)
Comment (sh-andriy) [reply addon_toolkit/interfaces/citation.py:43]: the **Mendeley** has a lot of different document types such as **book, journal** etc and our `ItemType.DOCUMENT` was not suitable for it to work properly that's why i made is as `str`.
When we with folks will be pushing Zotero's implementation we will discuss this also, i think we might change the `ItemType.DOCUMENT` logic a little

File: addon_imps/citations/mendeley.py (hunk 2 above)
Comment (sh-andriy) [reply addon_imps/citations/mendeley.py:22]: as I remember we agreed to not to apply paginations in citations for now, but I will ask about that to be sure

File: addon_imps/citations/mendeley.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                if filter_items is None or filter_items == item_type:
+                    items.append(
+                        ItemResult(
+                            item_id=item_id,
+                            item_name=item_name,
+                            item_type=item_type,
+                            item_path=None,
+                            csl=item_details.get("csl", {}),
+                        )
+                    )
```
[the remark this answers is not in the document]
Comment (sh-andriy) [reply addon_imps/citations/mendeley.py:69]: Good point, i think it is not necessary to place it there

File: addon_service/migrations/0002_configuredcitationaddon_root_folder.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+# Generated by Django 4.2.7 on 2024-08-20 13:45
```
[the remark this answers is not in the document]
Comment (sh-andriy) [reply addon_service/migrations/0002_configuredcitationaddon_root_folder.py:1]: We will be having discussion wi folks about this too, when we will be pushing Zotero implementation. There are some points to be discussed and this is one of them

File: addon_toolkit/tests/test_mendeley.py
Comment (opaduchak) [inline addon_toolkit/tests/test_mendeley.py:1]: I'd put it inside addon_imps package, to be consistent with current tests structure

Comment (sh-andriy) [reply addon_toolkit/tests/test_mendeley.py:1]: sure, will move that to `addon_imps/citations`

File: addon_imps/tests/test_mendeley.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+                item_name="Doc Title 2",
+                item_type=ItemType.DOCUMENT,
+                item_path=[],
+                csl={},
+            ),
+        ]
+
+        expected_result = ItemSampleResult(items=expected_items, total_count=2)
+
+        self.assertEqual(result, expected_result)
```
Comment (ly-mariia) [inline addon_imps/tests/test_mendeley.py:125]: Will there be any order issues?

**Review by ly-mariia — CHANGES_REQUESTED** [review]

File: addon_imps/citations/mendeley.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    ) -> list[ItemResult]:
+        items = []
+
+        tasks = [
+            self.fetch_item_details(doc["id"], filter_items, items)
+            for doc in document_ids
+        ]
+        await asyncio.gather(*tasks)
+
+        return items
```
Comment (opaduchak) [inline addon_imps/citations/mendeley.py:51]: please rewrite this. `asyncio.gather` returns results of all of it's coroutines in a tuple

File: addon_imps/citations/mendeley.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+    async def fetch_item_details(
```
Comment (opaduchak) [inline addon_imps/citations/mendeley.py:53]: why this method is public? Either make all utility methods protected or public

File: addon_imps/citations/mendeley.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+            if filter_items is None:
+                items.append(
+                    ItemResult(
+                        item_id=item_id,
+                        item_name=item_name,
+                        item_type=ItemType.DOCUMENT,
+                        item_path=item_details.get("path", []),
+                        csl=item_details.get("csl", {}),
+                    )
+                )
```
Comment (opaduchak) [inline addon_imps/citations/mendeley.py:69]: I am stricly opposed to mutating input arguments. Gladly this is done in async context

**Review by opaduchak — CHANGES_REQUESTED** [review]

File: addon_imps/tests/test_mendeley.py (hunk 7 above)
Comment (sh-andriy) [reply addon_imps/tests/test_mendeley.py:125]: Great point, solved that with my latest commit

File: addon_toolkit/interfaces/citation.py [hunk 11]
```diff
@@ -38,7 +38,7 @@ class ItemType(Enum):
 
 @dataclasses.dataclass(slots=True)
 class ItemResult:
-    item_id: int
+    item_id: str
```
[the remark this answers is not in the document]
Comment (sh-andriy) [reply addon_toolkit/interfaces/citation.py]: Yep, Zotero implementation has `str` here to , so we shouldn't have any problems

**Comment by brianjgeiger** [comment]
Looks like there's a merge conflict that needs to be handled.

**Review by ly-mariia — APPROVED** [review]

**Review by brianjgeiger — APPROVED** [review]

**Pull request merged**